internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,118 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(url, clientId, clientSecret, offlineToken string) *LazyToken {
+	return &LazyToken{
+		url:          url,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+		offlineToken: offlineToken,
+		token:        "",
+		expiration:   time.Now(),
+	}
+}
+
+// Token returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) Token(ctx context.Context) string {
+	lt.mutex.Lock()
+	defer lt.mutex.Unlock()
+
+	if lt.token == "" || lt.expiration.Before(time.Now().Add(time.Minute)) {
+
+		// Acquire a new token and set the expiration time
+		lt.token, _ = acquireToken(lt.url, lt.clientId, lt.clientSecret, lt.offlineToken)
+		lt.expiration = time.Now().Add(time.Hour)
+		logrus.WithContext(ctx).Infof("Acquired new token %s with expiration at: %s", lt.token, lt.expiration)
+	} else {
+		logrus.WithContext(ctx).Infof("Token requested: %s (expires at %s)", lt.token, lt.expiration)
+	}
+
+	return lt.token
+}
+
+// AcquireToken is a placeholder function that simulates token acquisition.
+func acquireToken(tokenURL, clientID, clientSecret, offlineToken string) (string, error) {
+	req, err := http.NewRequest("POST", tokenURL, nil)
+	if err != nil {
+		return "", err
+
+	}
+	data := url.Values{}
+
+	if offlineToken != "" {
+		data.Set("grant_type", "refresh_token")
+		data.Set("client_id", clientID)
+		data.Set("refresh_token", offlineToken)
+	} else if clientSecret != "" {
+		data.Set("grant_type", "client_credentials")
+		data.Set("client_id", clientID)
+		data.Set("client_secret", clientSecret)
+	}
+
+	req.Body = io.NopCloser(strings.NewReader(data.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		// Handle error
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		// Handle error response
+		fmt.Println("Error acquiring token:", string(body))
+		return "", err
+	}
+
+	var tokenResp tokenResponse
+	err = json.Unmarshal(body, &tokenResp)
+	if err != nil {
+		fmt.Println("Error unmarshal token response:", tokenResp)
+		return "", err
+	}
+
+	return tokenResp.AccessToken, nil
+}
```

Comment:
I suggest to update this function to return `tokenResponse` so you can work with the data in the caller.